docs/madr/decisions/039-transition-to-new-policies.md

```diff
@@ -0,0 +1,644 @@
+
+The goal is to provide a way to transition from old policies to new policies. Having a smooth transition is going to enable 
+us to deprecate old policies in the future.
+
+Prerequisites for the transition:
+* both Global and Zone CPs are running the latest Kuma version
+
+## Decision Drivers
+
+- Predictability: when replacement policies are created (manually or automatically), user should be able to verify them 
+without applying for real
+- Short feedback loop: we'd like to deliver small portions of new policies more often, rather than big portions but rarely
+- Quick rollback: user should be able to rollback to old policies
+
+## Considered Options
+
+- Introduce a shadow mode for policies on Zone CP
+- Introduce a shadow mode for policies on Global CP
+  - Show potential changes for a specific DPP
+  - Show potential changes for all affected DPPs
+
+## Decision Outcome
+
+- Introduce a shadow mode for policies only on Zone CP
+
+## Shadow mode for policies
+
+Shadow mode is a way to test new policies without affecting the real traffic. When the policy is created with `kuma.io/effect: shadow` label
+
+```yaml
+type: MeshTimeout
+name: timeout-global
+mesh: default
+labels:
+  kuma.io/effect: shadow
+spec:
+  targetRef:
+    kind: Mesh
+  to: [...]
+```
+
+it doesn't apply on the real proxies. Instead, by using Inspect API we can inspect what config dump would look like if the shadow policies was applied.
+
+### Changes in the Inspect API
+
+Introduce a new non-physical endpoint `/_snapshot` that returns a config dump of a DPP without the actual `/config_dump` call to Envoy.
+If we want to be able to compute diff on the client side we can't use `xds` endpoint as its response slightly diverges from what CP has in the SnapshotCache.
+
+Add a query parameter `shadow=true`:
+
+* `GET /meshes/{mesh}/dataplanes/{name}/_snapshot` -> `GET /meshes/{mesh}/dataplanes/{name}/_snapshot?shadow=true`
+* `GET /meshes/{mesh}/dataplanes/{name}/_rules` -> `GET /meshes/{mesh}/dataplanes/{name}/_rules?shadow=true`
+
+when `shadow=true` is set, the response will take policies with `kuma.io/effect: shadow` label into account.
+
+For the debugging purposes it makes sense to represent the shadow snapshot as the original snapshot + jsonPatch that can be applied to it.
+Kuma GUI and kumactl also can benefit from such representation. To make server produce such response we can introduce a query parameter `include=diff`.
+Query param `include` contains a list of fields that should be included in the response. By default, only `xds` field is set by the API server. 
+From the validation point of view the request `_snapshot?shadow=true&include=diff` makes no sense and should be rejected.  
+
+OpenAPI spec for new endpoints:
+
+```yaml
+openapi: 3.0.3
+info:
+  title: Shadow Policies Inspector
+  description: API for inspecting policies' effect on the Data Plane Proxy configurations
+  version: 1.0.0
+paths:
+  /meshes/{mesh}/dataplanes/{name}/_snapshot:
```

Comment:
Is `_snapshot` just to generic? Should this just be `_config` ? Istio uses `proxy-config` https://istio.io/latest/docs/ops/diagnostic-tools/proxy-cmd/

Reply:
I called it "snapshot" because CP operates with Snapshots and stores them in the SnapshotCache. This is contrasted to the config dump (or proxy-config) that is supposed to be taken from the proxy (not from CP). 

But if we want to converge `_snapshot` and `xds` in the future and introduce something like `mode=proxy|shadow` then `_config` is a better name

Reply:
I'd go for _config because I find _snapshot a lot more confusing (makes me think about DB snapshot as well).